File: one_time_scripts/parse_report.rb
#!/usr/bin/env ruby

require 'date'
require 'mongoid'


# Description: A one-off script that parses the police report and
# inserts them into a MongoDB db in a collection called "incidents"
#
# Notes:
#
# It does NOT fetch the page from the website:
#
# http://www.quincyma.gov/Government/POLICE/HotSpots.cfm
#
# ... so you would have to copy and paste the text from the site, and
# sift through the incidents you want to save.
#
# I know, I know... pretty lame, but the site isn't updated often anyways.
# If this script ever needs to called often, then update it so.


Mongoid.load!('./mongoid.yml', :development)
class Incident
  include Mongoid::Document

  field :id, type: Integer
  field :date, type: Date
  field :type, type: String
  field :recorded, type: String
  field :callTaker, type: String
  field :disposition, type: String
  field :officer, type: String
  field :notes, type: String

end



file = File.read("./sample.txt")
incidents = []
file.split(/^[\*_]+$/).each do |incident|
    incident = incident.strip()
    if !incident.start_with? "Incident"
        next
    end

    id = incident.match(/^Incident #: (\d+)/)[1]
    date = incident.match(/Date: (\d{4})-(\d{1,2})-(\d{1,2}) (\d{2}):(\d{2}):(\d{2})/)
    type = incident.match(/Type: (.*)$/)[1]
    location = incident.match(/^Location: (.*)$/)[1]
    howRec = incident.match(/How Rec: ([\w ]*) Call Taker/).nil? ? "N/A" : incident.match(/How Rec: (\w*) Call Taker/)[1]
    callTaker = incident.match(/Call Taker: (\w*)/)[1]
    disposition = incident.match(/^Disposition: ([\w ]*) Officer:/)[1]
    officer = incident.match(/Officer: ([\w, ]*)$/).nil? ? "<null>" : incident.match(/Officer: ([\w, ]*)$/)[1]
    notes = incident.split("Notes:\n")[1]

    incidents << {
        id: id.to_i,
        date:  DateTime.new(date[1].to_i, date[2].to_i, date[3].to_i, date[4].to_i, date[5].to_i, date[6].to_i).mongoize,
        type: type,
        location: location,
        recorded: howRec,
        callTaker: callTaker,
        disposition: disposition,
        officer: officer,
        notes: notes.nil? ? "" : notes.strip()
    }

end

Incident.collection.insert(incidents)

File: one_time_scripts/parse_report.js
/*global require,console*/

/**
 * A script that parses the police logs and stores them into a Postgresql database.
 */
(function () {
    'use strict';

    var fs = require('fs'),
        pg = require('pg'),
        Q = require('q');

    fs.readFile('./sample.txt', 'utf8', function (err, data) {

        var incidents = [],
            client;

        if (err) {
            return console.log(err);
        }

        // foreach block of text between a bunch of asterisks (*) and underscores (_)
        data.split(/[\*_]+/gm).forEach(function (incidentData) {
            var incident = {}, dateRegex;

            // trim it just in case
            incidentData = incidentData.trim();

            // incidents start with "Incident"
            if (incidentData.indexOf("Incident") === 0) {
                incident.id = incidentData.match(/^Incident #: (\d+)/)[1];
                incident.type = incidentData.match(/Type: (.*)$/m)[1];
                incident.location = incidentData.match(/^Location: (.*)$/m)[1];
                incident.recorded = incidentData.match(/How Rec: ([\w ]*) Call Taker/) ? incidentData.match(/How Rec: (\w*) Call Taker/)[1] : "N/A";
                incident.callTaker = incidentData.match(/Call Taker: (\w*)/)[1];
                incident.disposition = incidentData.match(/^Disposition: ([\w ]*) Officer:/m)[1];
                incident.officer = incidentData.match(/Officer: ([\w, ]*)$/) ? incidentData.match(/Officer: ([\w, ]*)$/)[1] : null;
                incident.notes = incidentData.split("Notes:\n")[1];

                // date needs to parsed
                dateRegex = incidentData.match(/Date: (\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})/);
                incident.date = new Date(parseInt(dateRegex[1], 10),
                    parseInt(dateRegex[2], 10) - 1, // javascript's Date's month is 0-based.  idunno, it just is.
                    parseInt(dateRegex[3], 10),
                    parseInt(dateRegex[4], 10),
                    parseInt(dateRegex[5], 10),
                    parseInt(dateRegex[6], 10));

                incidents.push(incident);
            }
        });

        // make `subType` from `type`
        incidents.forEach(function (incident) {
            incident.subType = incident.type.match(/\(([\w \/]+)\)/)[1];
        });


        client = new pg.Client('postgres://localhost/hotspots');
        client.connect(function (err) {

            if (err) {
                return console.log('Error connecting to postgres!');
            }

            var queue = Q.delay(1);
            incidents.forEach(function (incident) {
                var insert_statement_sql = 'INSERT INTO incident (id,date,type,location,recorded,call_taker,disposition,officer,notes) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)';
                queue = queue.then(function () {
                    client.query(insert_statement_sql, [
                        incident.id,
                        incident.date,
                        incident.type,
                        incident.location,
                        incident.recorded,
                        incident.callTaker,
                        incident.disposition,
                        incident.officer,
                        incident.notes
                    ], function (err) {
                        if (!err) {
                            console.log('Successfully added incident #: ' + incident.id);
                        } else {
                            console.log('Error adding incident #: ' + incident.id);
                        }
                    });
                }).delay(10);
            });

            queue = queue.done(function () {
                client.end();
            });
        });


    });

}());